### PR TITLE
[FLINK-15858][hive] Store generic table schema as properties

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

Store schema of generic table as properties, so that `HiveCatalog` is able to support types that are not supported by Hive, e.g. `TIMESTAMP(!=9)`.

## Brief change log

  - Store table schema as properties when creating generic tables
  - Retrieve table schema from properties when getting generic tables
  - Add test

## Verifying this change

Existing and new test cases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by lirui-apache** [comment]
cc @JingsongLi @bowenli86

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 1]
```diff
@@ -587,57 +599,67 @@ protected static Table instantiateHiveTable(ObjectPath tablePath, CatalogBaseTab
...
-				List<FieldSchema> partitionColumns = allColumns.subList(allColumns.size() - partitionKeySize, allColumns.size());
+		if (isGeneric) {
+			DescriptorProperties tableSchemaProps = new DescriptorProperties(true);
+			tableSchemaProps.putTableSchema(HiveCatalogConfig.GENERIC_TABLE_SCHEMA_PREFIX, table.getSchema());
+			properties.putAll(tableSchemaProps.asMap());
+
+			if (table instanceof CatalogTableImpl) {
+				List<String> partColNames = ((CatalogTableImpl) table).getPartitionKeys();
+				if (!partColNames.isEmpty()) {
+					properties.put(HiveCatalogConfig.GENERIC_PART_COL_NAMES, String.join(",", partColNames));
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: For compatibility issue, should throw exception now.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogConfig.java [hunk 2]
```diff
@@ -32,4 +32,9 @@
 
 	// Partition related configs
 	public static final String PARTITION_LOCATION = "partition.location";
+
+	// config prefix for the table schema of a generic table
+	public static final String GENERIC_TABLE_SCHEMA_PREFIX = "generic.table.schema";
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogConfig.java]: consistent with `CatalogConfig.FLINK_PROPERTY_PREFIX`?

**Comment by JingsongLi** [comment]
Can hive's table have no fields?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogConfig.java (hunk 2 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogConfig.java]: `CatalogConfig.FLINK_PROPERTY_PREFIX` will be added to all the Flink properties. So we'll have sth like `flink.generic.table.schema.0...` and `flink.generic.table.schema.1...` in the final table properties.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java (hunk 1 above)
Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: We don't need support partition for generic table. And this serialization maybe not right.
We should wait https://github.com/apache/flink/pull/10059

**Comment by lirui-apache** [comment]
PR updated to disallow creating partitioned generic table with HiveCatalog.

**Comment by lirui-apache** [comment]
> Can hive's table have no fields?

I thought not. But it turned out hive doesn't complain about that. So I think we're good.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java (hunk 1 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: OK I'll create another JIRA to track that.

**Comment by danny0405** [comment]
What is a 'generic table', it seems not a  jargon or terminology

**Comment by lirui-apache** [comment]
> What is a 'generic table', it seems not a jargon or terminology

HiveCatalog can handle two kinds of tables: hive-compatible tables and generic tables. For the former, we store the metadata in a hive-compatible way so that Hive itself can access those tables. For the latter, we just use HMS as a storage to persist the metadata, and Hive won't be able to (and doesn't need to) understand that metadata. I guess the latter one is `generic` from HMS's point of view.

**Comment by lirui-apache** [comment]
My personal build succeeds: https://travis-ci.org/lirui-apache/flink/builds/645851575

**Comment by danny0405** [comment]
> > What is a 'generic table', it seems not a jargon or terminology
> 
> HiveCatalog can handle two kinds of tables: hive-compatible tables and generic tables. For the former, we store the metadata in a hive-compatible way so that Hive itself can access those tables. For the latter, we just use HMS as a storage to persist the metadata, and Hive won't be able to (and doesn't need to) understand that metadata. I guess the latter one is `generic` from HMS's point of view.

Thanks for the explanation !

**Comment by bowenli86** [comment]
please see my comments in the jira

**Comment by lirui-apache** [comment]
I don't think the python failure is related. It has been failing for several other PRs lately.

**Pull request merged**